infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java

```diff
@@ -73,6 +100,83 @@ public boolean isPredefined() {
     return isPredefined;
   }
 
+  /** Returns list of k8s namespaces names where a user is able to run workspaces. */
+  public List<KubernetesNamespaceMeta> list() throws InfrastructureException {
+    if (!allowUserDefinedNamespaces) {
+      // the default namespace must be configured if user defined are not allowed
+      // so return only it
+      String evaluatedName =
+          evalDefaultNamespaceName(
+              defaultNamespaceName, EnvironmentContext.getCurrent().getSubject());
+
+      Namespace namespace;
+      try {
+        namespace = clientFactory.create().namespaces().withName(evaluatedName).get();
+      } catch (KubernetesClientException e) {
+        throw new InfrastructureException(
+            "Error occurred when tried to fetch default namespace. Cause: " + e.getMessage(), e);
+      }
+
+      KubernetesNamespaceMeta defaultNamespace;
+      if (namespace == null) {
+        // if the predefined namespace does not exist - return dummy info and it will be created
+        // during the first workspace start
+        defaultNamespace = new KubernetesNamespaceMetaImpl(evaluatedName);
+      } else {
+        defaultNamespace = asNamespaceMeta(namespace);
+      }
+
+      defaultNamespace.getAttributes().put(DEFAULT_ATTRIBUTE, "true");
+      return singletonList(defaultNamespace);
+    }
+
+    // if user defined namespaces are allowed - fetch all available
+    List<KubernetesNamespaceMeta> namespaces;
+    try {
+      namespaces =
+          clientFactory
+              .create()
+              .namespaces()
+              .list()
+              .getItems()
+              .stream()
+              .map(this::asNamespaceMeta)
+              .collect(Collectors.toList());
+    } catch (KubernetesClientException e) {
+      throw new InfrastructureException(
+          "Error occurred when tried to list all available namespaces. Cause: " + e.getMessage(),
+          e);
+    }
+
+    // propagate default namespace if it's configured
+    if (!isNullOrEmpty(defaultNamespaceName)) {
+      String evaluatedName =
+          evalDefaultNamespaceName(
+              defaultNamespaceName, EnvironmentContext.getCurrent().getSubject());
+
+      Optional<KubernetesNamespaceMeta> defaultNamespaceOpt =
+          namespaces.stream().filter(n -> evaluatedName.equals(n.getName())).findAny();
+      KubernetesNamespaceMeta defaultNamespace;
+      if (defaultNamespaceOpt.isPresent()) {
+        defaultNamespace = defaultNamespaceOpt.get();
+      } else {
+        defaultNamespace = new KubernetesNamespaceMetaImpl(evaluatedName);
+        namespaces.add(defaultNamespace);
+      }
+
+      defaultNamespace.getAttributes().put(DEFAULT_ATTRIBUTE, "true");
+    }
+    return namespaces;
+  }
+
+  private KubernetesNamespaceMeta asNamespaceMeta(Namespace namespace) {
+    Map<String, String> attributes = new HashMap<>();
```

Comment:
A bit of a nitpick but we're creating a hashmap with preinitialized capacity of (at least) 16 to put at most 1 element there.

Reply:
Since `DEFAULT_LOAD_FACTOR` is 0.75 - two `new HashMap<>(2);` should be enough. Are you OK with that?